File: apps/api/internal/domain/metrics/metrics_entity.go
package metrics

import (
	"errors"
	"time"
)

// ErrNotFound is returned when a metric is not found.
var ErrNotFound = errors.New("metric not found")

// TelemetryFrame represents a raw telemetry data point from a device.
type TelemetryFrame struct {
	Timestamp   time.Time `json:"timestamp"`
	DeviceID    string    `json:"deviceId"`
	RiskScore   float64   `json:"riskScore"`
	ThermalTemp float64   `json:"thermalTemp"`
	BufferLevel float64   `json:"bufferLevel"`
	Uptime      int64     `json:"uptime"`
}

// MetricDataPoint represents an aggregated metric data point for charts.
type MetricDataPoint struct {
	Timestamp time.Time `json:"timestamp"`
	Value     float64   `json:"value"`
}

// MetricStats represents statistical summary of a metric.
type MetricStats struct {
	Current float64 `json:"current"`
	Avg     float64 `json:"avg"`
	Min     float64 `json:"min"`
	Max     float64 `json:"max"`
}

// MetricThreshold represents warning and critical thresholds.
type MetricThreshold struct {
	Warning  float64 `json:"warning"`
	Critical float64 `json:"critical"`
}

// MetricData represents a complete metric with stats, chart data, and thresholds.
type MetricData struct {
	Unit      string            `json:"unit"`
	Chart     []MetricDataPoint `json:"chart"`
	Stats     MetricStats       `json:"stats"`
	Threshold MetricThreshold   `json:"threshold"`
}

// DeviceMetrics represents all metrics for a device.
type DeviceMetrics struct {
	TimeRange TimeRangeInfo          `json:"timeRange"`
	Device    DeviceInfo             `json:"device"`
	Events    []MetricThresholdEvent `json:"events,omitempty"`
	Metrics   MetricsCollection      `json:"metrics"`
}

// DeviceInfo contains device identification.
type DeviceInfo struct {
	IMEI       string `json:"imei"`
	DeviceName string `json:"deviceName,omitempty"`
}

// TimeRangeInfo contains the time range parameters.
type TimeRangeInfo struct {
	Start      time.Time `json:"start"`
	End        time.Time `json:"end"`
	Range      string    `json:"range"`
	Resolution string    `json:"resolution"`
}

// MetricsCollection contains all metric types.
type MetricsCollection struct {
	RiskScore   MetricData `json:"riskScore"`
	ThermalTemp MetricData `json:"thermalTemp"`
	BufferLevel MetricData `json:"bufferLevel"`
	Uptime      MetricData `json:"uptime"`
}

// MetricThresholdEvent represents an event when a threshold is breached.
type MetricThresholdEvent struct {
	Timestamp time.Time `json:"timestamp"`
	Type      string    `json:"type"`
	Metric    string    `json:"metric"`
	Value     float64   `json:"value"`
	Threshold float64   `json:"threshold"`
}

// ThresholdPreset represents operator-specific threshold settings.
type ThresholdPreset struct {
	RiskScoreWarning  float64 `json:"riskScoreWarning"`
	RiskScoreCritical float64 `json:"riskScoreCritical"`
	ThermalWarning    float64 `json:"thermalWarning"`
	ThermalCritical   float64 `json:"thermalCritical"`
	BufferWarning     float64 `json:"bufferWarning"`
	BufferCritical    float64 `json:"bufferCritical"`
}

File: packages/API_Client/src/domain/metrics/metrics-entity.ts
export type TimeRange = "1h" | "6h" | "24h" | "7d";

export type MetricResolution = "1m" | "5m" | "15m" | "1h";

export interface MetricChartPoint {
  timestamp: Date;
  value: number;
}

export interface MetricThreshold {
  warning: number;
  critical: number;
}

export interface MetricData {
  current: number;
  avg: number;
  min: number;
  max: number;
  unit: string;
  chart: MetricChartPoint[];
  threshold: MetricThreshold;
}

export interface UptimeData {
  current: number;
  unit: string;
}

export interface MetricEvents {
  timestamp: Date;
  type: string;
  metric: string;
  value: number;
  threshold: number;
}

export interface TimeRangeInfo {
  start: Date;
  end: Date;
  range: TimeRange;
  resolution: MetricResolution;
}

export interface DeviceBasic {
  imei: string;
  deviceName?: string;
}

export interface DeviceMetrics {
  device: DeviceBasic;
  timeRange: TimeRangeInfo;
  metrics: {
    riskScore: MetricData;
    thermalTemp: MetricData;
    bufferLevel: MetricData;
    uptime: UptimeData;
  };
  events: MetricEvents[];
}

export interface DashboardStats {
  devices: {
    total: number;
    online: number;
    offline: number;
  };
  commands: {
    totalToday: number;
    pending: number;
    failed: number;
  };
  activity: {
    last24h: {
      commands: number;
      registrations: number;
      deregistrations: number;
    };
  };
}

export function getTimeRangeMs(range: TimeRange): number {
  const ranges: Record<TimeRange, number> = {
    "1h": 60 * 60 * 1000,
    "6h": 6 * 60 * 60 * 1000,
    "24h": 24 * 60 * 60 * 1000,
    "7d": 7 * 24 * 60 * 60 * 1000,
  };
  return ranges[range];
}

export function getResolutionForRange(range: TimeRange): MetricResolution {
  const resolutions: Record<TimeRange, MetricResolution> = {
    "1h": "1m",
    "6h": "5m",
    "24h": "15m",
    "7d": "1h",
  };
  return resolutions[range];
}

export function getTimeRangeLabel(range: TimeRange): string {
  const labels: Record<TimeRange, string> = {
    "1h": "1 Hour",
    "6h": "6 Hours",
    "24h": "24 Hours",
    "7d": "7 Days",
  };
  return labels[range];
}
